flips/20220715-cadence-purity-analysis.md

```diff
@@ -0,0 +1,211 @@
+# Function Purity 
+
+| Status        | (Proposed)       |
+:-------------- |:---------------------------------------------------- |
+| **FLIP #**    | [1056](https://github.com/onflow/flow/pull/1056)     |
+| **Author(s)** | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Sponsor**   | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Updated**   | 2022-08-25                                           |
+
+## Objective
+
+This FLIP proposes to add a new syntax and accompanying semantic analysis to determine or 
+enforce that a given function or method is "view", that is, lacking in side effects. We would
+then add an additional checker rule to enforce purity in function conditions. 
+
+## Motivation and User Benefit
+
+An [issue](https://github.com/onflow/cadence/issues/1805) with function pre- and post-conditions allows arbitrary functions to be
+called in these condition blocks, which can allow implementers to write underhanded code by modifying state inside of a condition. 
+By allowing users to annotate their functions as `view`, we can enforce that these functions have no side effects that modify state
+and thus permit these functions (and only these functions) to be called from within a condition. 
+
+## Design Proposal
+
+### Parsing
+
+This adds a new keyword to the langauge: `view`. This is optionally placed before the `fun` keyword, but after the access modifier, 
+in function declarations and expressions. So, for example, all of the following are valid:
+
+```cadence
+pub view fun foo(): Void {}
+```
+
+```cadence
+view fun foo(): Void {}
+```
+
+```cadence
+let x = view fun(): Void {}
+```
+
+```cadence
+pub struct S {
+    pub view fun foo(): Void {}
+    view init()
+}
+```
+
+```cadence
+pub resource interface R {
+    priv view fun foo(): Void {}
+    view init()
+}
+```
+
+Any function that does not have a `view` annotation will be considered impure. 
+
+Function types can also have purity annotations, to be placed after the opening parenthesis but before the parameter list. So, for example, these are valid
+types:
```

Comment:
Like for the name of the proposal/feature, maybe use "view" consistently instead of pure/impure/purity annotations, here, and in the rest of the proposal:

```suggestion
Any function that does not have a `view` annotation will be considered a non-view function. 

Function types can also have a `view` annotations, to be placed after the opening parenthesis but before the parameter list. So, for example, these are valid types:
```